internal/controllers/user_integrations_controller.go

```diff
@@ -1933,6 +1950,49 @@ func (udc *UserDevicesController) registerDeviceTesla(c *fiber.Ctx, logger *zero
 	return c.SendStatus(fiber.StatusNoContent)
 }
 
+func (udc *UserDevicesController) getTeslaAuthFromCache(ctx context.Context, userID string) (*RegisterDeviceIntegrationRequest, error) {
+	user, err := udc.usersClient.GetUser(ctx, &pb.GetUserRequest{Id: userID})
+	if err != nil {
+		return nil, errors.Wrap(err, "could not fetch user information")
+	}
+
+	if user.EthereumAddress == nil {
+		return nil, fmt.Errorf("missing wallet address for user")
+	}
+
+	cacheKey := fmt.Sprintf(TeslaFleetAuthCacheKey, *user.EthereumAddress)
+	encTeslaAuth, err := udc.redisCache.Get(ctx, cacheKey).Result()
+	if err != nil {
+		udc.log.Err(err).Str("Cache Key", cacheKey).Msg("Error occurred retrieving tesla auth from cache using the key")
+		return nil, errors.Wrap(err, "an error occurred completing tesla device registration")
+	}
+	if len(encTeslaAuth) == 0 {
+		return nil, fmt.Errorf("an error occurred completing tesla device registration")
+	}
+	decrypted, err := udc.cipher.Decrypt(encTeslaAuth)
+	if err != nil {
+		udc.log.Err(err).Str("cacheKey", cacheKey).Msg("failed to decrypt tesla auth from cache using key")
+		return nil, errors.Wrap(err, "failed to decrypt tesla token")
+	}
+
+	teslaAuth := &services.TeslaAuthCodeResponse{}
+	err = json.Unmarshal([]byte(decrypted), &teslaAuth)
+	if err != nil {
+		udc.log.Err(err).Msgf("failed to unmarshal tesla token found in redis using cacheKey: %s", cacheKey)
+		return nil, errors.Wrap(err, "failed to parse tesla authorization token")
+	}
+
+	if teslaAuth == nil {
+		return nil, fmt.Errorf("could not find any tesla authorization for user")
+	}
+
+	return &RegisterDeviceIntegrationRequest{
+		AccessToken:  teslaAuth.AccessToken,
+		ExpiresIn:    int(teslaAuth.Expiry.Unix()),
```

Comment:
We keep mixing up expiry and "expires in". The first is a time, the second is a duration.